// Copyright 2018 Google LLC
// Copyright 2018-present Open Networking Foundation
// SPDX-License-Identifier: Apache-2.0


// The P4PipelineConfig contains data generated by the Stratum p4c backend
// to help Stratum switches intepret P4 runtime requests and P4Info content.

syntax = "proto3";

option cc_generic_services = false;

package stratum.hal;

import "p4/v1/p4runtime.proto";
import "stratum/hal/lib/p4/p4_control.proto";
import "stratum/hal/lib/p4/p4_table_map.proto";
import "stratum/public/proto/p4_annotation.proto";

// The P4PipelineConfig message conists of these fields:
//  table_map - contains a map from P4 object name to descriptor data that
//      supports mapping between P4 runtime requests, P4Info, and internal
//      HAL flow entries.
//  p4_controls - present the control methods from the P4 program in a form
//      that the HAL switch can use for logical-to-physical table translation.
//  idle_pipeline_stages - identify any forwarding pipeline stages that are
//      not needed by tables in the P4 program.
//  static_table_entries - contains a WriteRequest.updates() entry for each
//      "const entry" table property in the P4 program.
message P4PipelineConfig {
  map<string, P4TableMapValue> table_map = 1;
  repeated P4Control p4_controls = 2;
  repeated P4Annotation.PipelineStage idle_pipeline_stages = 3;
  p4.v1.WriteRequest static_table_entries = 4;
}
